internal/syscall/src/lib.rs

```diff
@@ -658,4 +678,207 @@ pub trait SyscallHandler: AddressValidator + Sized {
         self.trace("close", 1);
         unsafe { self.proxy(request!(libc::SYS_close => fd)) }
     }
+
+    /// syscall
+    fn socket(&mut self, domain: libc::c_int, type_: libc::c_int, protocol: libc::c_int) -> Result {
+        self.trace("socket", 3);
+        unsafe { self.proxy(request!(libc::SYS_socket => domain, type_, protocol)) }
+    }
+
+    /// syscall
+    fn bind(&mut self, fd: libc::c_int, addr: UntrustedRef<u8>, addrlen: libc::size_t) -> Result {
+        self.trace("bind", 3);
+        if addrlen > Block::buf_capacity() {
+            return Err(libc::EINVAL);
+        }
+
+        let addr = addr.validate_slice(addrlen, self).ok_or(libc::EFAULT)?;
+
+        let c = self.new_cursor();
+        let (_, addr) = c.copy_from_slice(addr.as_ref()).or(Err(libc::EMSGSIZE))?;
+        let addr = addr.as_ptr();
+        let host_virt = Self::translate_shim_to_host_addr(addr);
+
+        unsafe { self.proxy(request!(libc::SYS_bind => fd, host_virt, addrlen)) }
+    }
+
+    /// syscall
+    fn connect(
+        &mut self,
+        fd: libc::c_int,
+        addr: UntrustedRef<u8>,
+        addrlen: libc::size_t,
+    ) -> Result {
+        self.trace("connect", 3);
+        if addrlen > Block::buf_capacity() {
+            return Err(libc::EINVAL);
+        }
+
+        let addr = addr.validate_slice(addrlen, self).ok_or(libc::EFAULT)?;
+
+        let c = self.new_cursor();
+        let (_, addr) = c.copy_from_slice(addr.as_ref()).or(Err(libc::EMSGSIZE))?;
+        let addr = addr.as_ptr();
+        let host_virt = Self::translate_shim_to_host_addr(addr);
+
+        unsafe { self.proxy(request!(libc::SYS_connect => fd, host_virt, addrlen)) }
+    }
+
+    /// syscall
+    fn accept(
+        &mut self,
+        fd: libc::c_int,
+        addr: UntrustedRefMut<u8>,
+        addrlen: UntrustedRefMut<libc::socklen_t>,
+    ) -> Result {
+        self.accept4(fd, addr, addrlen, 0)
+    }
+
+    /// syscall
+    fn accept4(
+        &mut self,
+        fd: libc::c_int,
+        addr: UntrustedRefMut<u8>,
+        addrlen: UntrustedRefMut<libc::socklen_t>,
+        flags: libc::c_int,
+    ) -> Result {
+        self.trace("accept4", 4);
+
+        if addr.as_ptr().is_null() {
+            return unsafe {
+                self.proxy(
+                    request!(libc::SYS_accept4 => fd, addr.as_ptr(), addrlen.as_ptr(), flags),
```

Comment:
The man page suggests that if `addr` is passed in as null, then a null pointer should be passed in for `addrlen`.

Reply:
@connorkuehl  And? I just pass on what userspace gave me and userspace will get the errors it deserves.